pallets/protos/src/tests.rs

```diff
@@ -1,385 +1,885 @@
-	let data = DATA.as_bytes().to_vec();
-	let references = vec![];
-
-	assert_ok!(ProtosPallet::upload(
-		Origin::signed(sp_core::ed25519::Public::from_raw(PUBLIC1)),
-		references,
-		Categories::Text(TextCategories::Plain),
-		Vec::new(),
-		None,
-		None,
-		data,
-	));
-}
+mod upload_tests {
+	use super::*;
+
+	pub fn upload(
+		signer: <Test as frame_system::Config>::AccountId,
+		proto: &ProtoFragment,
+	) -> DispatchResult {
+		ProtosPallet::upload(
+			Origin::signed(signer),
+			proto.references.clone(),
+			proto.category.clone(),
+			proto.tags.clone(),
+			proto.linked_asset.clone(),
+			proto.include_cost.map(|cost| Compact::from(cost)),
+			proto.data.clone(),
+		)
+	}
+
+	#[test]
+	fn upload_should_work() {
+		new_test_ext().execute_with(|| {
+			let dd = DummyData::new();
+
+			let block_number = System::block_number(); //@sinkingsugar
+
+			let proto = dd.proto_fragment;
+
+			assert_ok!(upload(dd.account_id, &proto));
+
+			assert!(<Protos<Test>>::contains_key(proto.get_proto_hash()));
+
+			let proto_struct = <Protos<Test>>::get(proto.get_proto_hash()).unwrap();
+
+			// I am using `match` to ensure that this test case fails if a new field is ever added to the `Proto` struct
+			// match proto_struct {
+			// 	Proto {
+			// 		block: 1,
+			// 		patches: Vec::new(),
+			// 		include_cost: proto.include_cost.map(|cost| Compact::from(cost)),
+			// 		creator: dd.account_id,
+			// 		owner: dd.account_id,
+			// 		references: proto.references,
+			// 		category: proto.category,
+			// 		tags: proto.tags,
+			// 		metadata: BTreeMap::new(),
+			// 	} => (),
+			// 	// _ => println!("Time to panic!!! Mayday"),
+			// }
+
+			let correct_proto_struct = Proto {
+				block: block_number,
+				patches: Vec::new(),
+				include_cost: proto.include_cost.map(|cost| Compact::from(cost)),
+				creator: dd.account_id,
+				owner: ProtoOwner::User(dd.account_id),
+				references: proto.references.clone(),
+				category: proto.category.clone(),
+				tags: Vec::new(), // proto.tags,
+				metadata: BTreeMap::new(),
+				accounts_info: AccountsInfo::default(),
+			};
+
+			// Ensure that this test case fails if a new field is ever added to the `Proto` struct
+			match proto_struct {
+				s if s == correct_proto_struct => (),
+				_ => panic!("The correct `Proto` struct was not saved in the StorageMap `Protos`"),
+			}
+
+			assert!(<ProtosByCategory<Test>>::get(&proto.category)
+				.unwrap()
+				.contains(&proto.get_proto_hash()));
+			assert!(<ProtosByOwner<Test>>::get(ProtoOwner::User(dd.account_id))
+				.unwrap()
+				.contains(&proto.get_proto_hash()));
+
+			let event = <frame_system::Pallet<Test>>::events()
+				.pop()
+				.expect("Expected at least one EventRecord to be found")
+				.event;
+			assert_eq!(
+				event,
+				mock::Event::from(pallet_protos::Event::Uploaded {
+					proto_hash: proto.get_proto_hash(),
+					cid: proto.get_proto_cid()
+				})
+			);
+		});
+	}
+
+	#[test]
+	fn upload_should_not_works_if_proto_hash_exists() {
```

Comment:
Btw can you fix those names too? With proper English. The previous engineer didn't really compute... "works" 😝